notifications/v1/client.go

```diff
@@ -6,90 +6,153 @@ import (
 
 	"cloud.google.com/go/pubsub"
 	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/background"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"go.uber.org/atomic"
 )
 
-// Client is a client for receiving SAMS notifications from a GCP Pub/Sub
-// subscription.
-type Client struct {
-	logger        log.Logger
-	subscription  *pubsub.Subscription
+type subscriber struct {
+	logger       log.Logger
+	handlers     SubscriberHandlers
+	subscription *pubsub.Subscription
+
+	// state indicates the state of workers, must be one of "idle", "started" or
+	// "stopped".
```

Comment:
No changes needed, but something to consider:

These 3x states to form a basic state machine? e.g. Only a subscriber in the "idle" state can call `Start()`,  and only in the "started" state can you call `Stop()`.

The doc comment here is good, since it calls out something important that isn't clear from the code. But it may perhaps be useful to add another couple of sentences for what those states _mean_, and what invariants we expect to maintain.

Given the simplicity/clarify of this, updating the comment wouldn't be adding any real value. But you can see if there was a fourth state, that some other type of abstraction or clarification about that underlying state machine would be important.

Reply:
Actually pushed a commit to make such simple state machine https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/cd459a3a207d91257f2567979583f82de3f593f3 😂 (I failed the computation class that teaches the state machine so... LMK WDYT)